### PR TITLE
[Fixes #2] Add ability to store and view image links.

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [ ] The landing page links to the image link submission form.
- [ ] The link is entered through a form.
- [ ] After the form is saved, the link is persisted in the database.
- [ ] After the form is submitted, the user is redirected to a page displaying
  the image.

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

### PR DESCRIPTION
**Pull request by awright022**
Closes #2

Create a table for the images table with a url field.  Allow creation and viewing of images created.  Enforce validation on urls being passed in.

Link to create an image is visible from the home page index.

### CODE REVIEW

**Comment by awright022** [comment]
Okay, I re-committed the above PR with atomic commits and without the extra scaffolding bloat.  @HaominZhang or @bboe did you want to look at it again?

File: app/views/home/index.html.erb [hunk 1]
```diff
@@ -1,2 +1,3 @@
 <h1>Welcome to the Image Sharer App!</h1>
-<p>by RentLinx by Appfolio by Wizards by Tony</p>
+<h4>by RentLinx by Appfolio by Wizards by Tony</h4>
+<p>Would you like to <%= link_to 'add a new image', new_image_path %>?</p>
```
Comment (bboe) [inline app/views/home/index.html.erb]: ⭐️  I like that you're actually linking to this new action. Many people neglect that.

File: config/routes.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :images
```
Comment (bboe) [inline config/routes.rb]: Give `rake routes` a run and see if you've only defined the routes that you expect.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get new_image_url
+    assert_response :success
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Prefer to use `:ok` instead of `:success`.

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -1,7 +1,26 @@
...
+
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+
+    respond_to do |format|
+      if @image.save
+        format.html { redirect_to '/', notice: 'Image was successfully created.' }
+      else
+        format.html { render :new }
```
Comment (bboe) [inline app/controllers/images_controller.rb]: Should this respond with a 200, ok status?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -5,4 +5,16 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     get new_image_url
     assert_response :success
   end
+
+  test 'should create image' do
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: What about the failure case?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -5,4 +5,16 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    assert_difference('Image.count') do
+      post images_url, params: { image: { url: sample_image.url } }
+    end
+
+    assert_redirected_to '/'
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Should we verify the flash message?

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -11,10 +11,18 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
       post images_url, params: { image: { url: sample_image.url } }
     end
 
-    assert_redirected_to '/'
+    assert_redirected_to image_url(Image.last)
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: ⭐️  awesome!

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -11,10 +11,18 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+  end
+
+  test 'should show image' do
+    image = sample_image
+    image.save
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: For ActiveRecord, please use the `!` version of methods if you aren't going to check the return value. That way things blow up if the save fails for any reason.

File: app/views/images/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<h1>Add a new Image</h1>
+
+<%= render 'form', image: @image %>
```
Comment (HaominZhang) [inline app/views/images/new.html.erb]: Seems `form` is a partial, do we wanna use `render partial:`, add pass in `locals`?

Comment (awright022) [inline app/views/images/new.html.erb]: Hey @HaominZhang !  Just talked with Bryce here side-by-side.  He thinks that there likely isn't any value to adding a flow test to this story at this point in time.  Does that sound okay by you?

**Comment by awright022** [comment]
@bboe @HaominZhang Okay, so the flow test was decided to be unimportant and I just pushed up the change to use the render partial: call (Thanks Haomin!).  Anything else or are we good to merge this bad boy?

File: app/models/image.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+class Image < ApplicationRecord
+  # url validation sourced from http://stackoverflow.com/a/24659037
+  validates :url, presence: true, format: { with: URI.regexp }
```
Comment (XanderStrike) [inline app/models/image.rb]: That stack overflow answer trips everyone up unfortunately. If only I had more downvotes to give. `URI.regexp` is designed for pulling URL-like things out of huge blobs of text, for auto-linking comments and blog posts and such. It doesn't do much to make sure the URL is valid though, consider these test cases which pass your validation:

* 'http'
* 'http://a' (feel free to laugh at github for getting this one wrong 😉 
* 'this is not a url -> http://www.google.com' (extra text in the field)
* 'wat://google.com'

There are lots of good URL test cases and regexes floating around online, you'll definitely want to beef up your test to include a list of good and bad URLs to test against whatever regex you find.

**Pull request merged**